llm/gpt-3/modeling.py

```diff
@@ -96,109 +96,72 @@ class MultiHeadAttention(nn.Layer):
     Cache = collections.namedtuple("Cache", ["k", "v"])
     StaticCache = collections.namedtuple("StaticCache", ["k", "v"])
 
-    def __init__(
-        self,
-        embed_dim,
-        num_heads,
-        dropout=0.0,
-        kdim=None,
-        vdim=None,
-        need_weights=False,
-        weight_attr=None,
-        output_layer_weight_attr=None,
-        bias_attr=None,
-        fuse_attention_qkv=False,
-        scale_qk_coeff=1.0,
-        num_partitions=1,
-        fused_linear=False,
-        use_recompute=False,
-        recompute_granularity="full",
-        do_recompute=True,
-        use_flash_attn=False,
-    ):
+    def __init__(self, config,):
         super(MultiHeadAttention, self).__init__()
-        self.embed_dim = embed_dim
-        self.kdim = kdim if kdim is not None else embed_dim
-        self.vdim = vdim if vdim is not None else embed_dim
-        self.num_heads = num_heads
-        self.dropout = dropout
-        self.need_weights = need_weights
-        self.fuse_attention_qkv = fuse_attention_qkv
-        self.scale_qk_coeff = scale_qk_coeff
-        self.use_recompute = use_recompute
-        self.recompute_granularity = recompute_granularity
-        self.do_recompute = do_recompute
-        self.use_flash_attn = use_flash_attn if flash_attention else None
-
-        self.head_dim = embed_dim // num_heads
-        assert self.head_dim * num_heads == self.embed_dim, "embed_dim must be divisible by num_heads"
-
-        assert self.num_heads % num_partitions == 0
-        self.num_heads = self.num_heads // num_partitions
-
-        if num_partitions > 1:
-            if self.fuse_attention_qkv:
-                assert self.kdim == embed_dim, "embed_dim should be equal to kdim"
-                assert self.vdim == embed_dim, "embed_dim should be equal to vidm"
 
+        self.config = config
+
+        # Recompute defaults to False and is controlled by Trainer
+        self.enable_recompute = False
+
+        config.use_flash_attention = config.use_flash_attention if flash_attention else None
+
+        self.head_dim = config.hidden_size // config.num_attention_heads
+        assert self.head_dim * config.num_attention_heads == config.hidden_size, "hidden_size must be divisible by num_attention_heads"
+
+        if config.tensor_parallel_degree > 1:
+            assert config.num_attention_heads % config.tensor_parallel_degree == 0
+            config.num_attention_heads = config.num_attention_heads // config.tensor_parallel_degree
```

Comment:
```suggestion
            self.num_attention_heads = config.num_attention_heads // config.tensor_parallel_degree
```
修改了原始变量的地方，建议重新赋值一遍。不要直接修改config